text/0000-local-snapshot-file-format/0000-local-snapshot-file-format.md

```diff
@@ -0,0 +1,186 @@
++ Feature name: `local_snapshot_file_format`
++ Start date: 2020-08-25
++ RFC PR: [iotaledger/protocol-rfcs#0000](https://github.com/iotaledger/protocol-rfcs/pull/0000)
+
+# Summary
+
+This RFC defines a file format for local snapshots which is compatible with Chrysalis Phase 2.
+
+# Motivation
+
+Nodes create local snapshots to produce ledger representations at a point in time of a given milestone in order to be able to:
+* start up from a recent milestone instead of having to synchronize from genesis
+* delete transaction data below the given milestone
+
+Current node implementations use a [local snapshot file format](https://github.com/iotaledger/iri-ls-sa-merger/tree/351020d3b5e342b6e9a41f2868575ab7ff8c251c#generating-an-export-file-from-a-localsnapshots-db) which only works with account based ledgers. For Chrysalis Phase 2 this file format has to be assimilated to support a UTXO based ledger.
+
+# Detailed design
+
+All types are serialized in little-endian and occur in the sequence of the rows defined below. Local snapshot files are compressed via zlib to further reduce size.
+
+SEP = solid entry point. `Array[T]` are prefixed with a varint denoting the length.
+
+This format describes version 1:
+<table>
+    <tr>
+        <th>Name</th>
+        <th>Type</th>
+        <th>Description</th>
+    </tr>
+    <tr>
+        <td>Version</td>
+        <td>byte</td>
+        <td>
+        The version of the local snapshot file format.
+        </td>
+    </tr>    
+    <tr>
+        <td>Milestone Index</td>
+        <td>uint64</td>
+        <td>
+        The index of the milestone of the local snapshot.
+        </td>
+    </tr>
+    <tr>
+        <td>Milestone Hash</td>
+        <td>ByteArray[32]</td>
+        <td>
+        The BLAKE2b-256 hash of the milestone payload.
+        </td>
+    </tr>
+    <tr>
+        <td>Timestamp</td>
+        <td>uint64</td>
+        <td>
+        The UNIX epoch timestamp in seconds of when this snapshot was created.
+        </td>
+    </tr>
+    <tr>
+        <td>SEPs</td>
+        <td>Array[ByteArray[32]]</td>
+        <td>
+        The BLAKE2b-256 hashes of the SEP messages at the cut off point of the given milestone.
+        </td>
+    </tr>
+    <tr>
+        <td>UTXOs</td>
+        <td colspan="2">
+            <details open="true">
+                <summary>Array[UTXO]</summary>
```

Comment:
If I understand correctly this array will be preceded by a varint specifying the element count of the Array itself; while forming the file you may not know ahead of time how many UTXOs to persist, and therefore the precise offset of the UTXOs array, shifted by its count varint. Seeking and moving on that file descriptior might be quite annoying: perhaps it would be better to detach your Array definition from its respective varint count, so you can manipulate the type of this latter depending on the circumstance. In this case you are probably better off with a fixed type for the UTXOs count.

Reply:
Yeah unfortunately I only wrote the export code after writing the RFC. I am currently writing up another format which also allows iterators for UTXOs.